### ISSUE TITLE
USDR Superuser has ability to update ARPA input template from the web application

### ISSUE THREAD

**Issue by as1729** [usdigitalresponse/usdr-gost#1267]
## Context
Updating the arpa input template is a common task that may need to happen at the end of a quarter. This is something that needs to be accomplished and deployed within a short period of time in order to unblock partners.

## Current State
Each time we need to update a template we'll need to make a PR similar to this #1264 . This is ends up making the CI/CD pipeline and engineering resources a dependency to getting partners an updated input template.

## Future State
When a new ARPA input template is available. An Admin/super-user can go into the ARPA tool and update the template version by uploading the new file.

## Design
@caitlinwinner 's team to confirm this-
There will likely be two new pages added.

1. Template form upload page
- User visits /arpa_reporter/update-template-version
- Page displays a form to upload a file
- When a file is uploaded, there is a view to confirm the new file version as well as the updated `templateRules.json`
- If the user is satisfied with the changes they can apply the new version for all partners.

2. Template form view page
- User can also visit a new page /arpa_reporter/current-template-version
- This page will display the current version name as well as the `templateRules.json` file for the admin to inspect any issues.

## Engineering
To Add
- `arpa-input-template-files` S3 bucket to hold the input template files
- `arpa-template-rules` S3 bucket to hold the `templateRules.json` files. Note: we will continue to hold all previous versions of this file.
- `arpa_templates` Database table to manage template records

DB Schema
`arpa_templates`
- `id` : pk int
- `external_id` : uuid that can be used while saving file to s3
- `input_filename` : varchar - points to a S3 key in `arpa-input-template-files` bucket
- `rules_filename` : varchar - points to a S3 Key `arpa-template-rules` bucket
- `effective_at`: datetime - the time at which point the file was made to be the latest arpa input template

```sql
/* to retrieve the latest arpa template */
SELECT      *
FROM         arpa_templates
ORDER BY effective_at DESC
LIMIT          1;
```

To Update
- Code that refers to `templateRules.json` will now need to retrieve this data from S3
- Code that refers to the input templates will now need to retrieve it from S3.

New Endpoints
- POST /upload-new-template-version
  -  Validates that user is bumping the version up and checks that the name follows known format.
  - Adds file to `template-files` S3 bucket
  - Adds a record in `arpa_templates` with the current_time as the `effective_at`
  - Runs the same script as `yarn workspace server run generate-rules` to update `templateRules.json` and upload the file to S3.
  - Redirects to`/arpa-input-template-version/latest` upon success.

- GET /arpa-input-template-version/:versionId
  - If a :versionId is not supplied default to `latest` or most effective version of the template
  - Returns `{ inputTemplate: <signed S3 URL to download template>, templateRules: <JSON from S3> }`

**Comment by as1729** [comment]
@jcomeau1 @NatHillardUSDR @TylerHendrickson as discussed in one of the slack threads we would like to "automate" as much as possible the process to update an ARPA input template. I thought it made sense for this to be a flow directly within the app that @jcomeau1 or someone else could manage. Let me know what you think.

**Comment by NatHillardUSDR** [comment]
This is a great idea! We should also do this for the output templates

**Comment by TylerHendrickson** [comment]
Consideration to discuss: We may be able to get some automation in place sooner if we developed a CI workflow (e.g. using GitHub actions) that did one of the following:
- Reacts to a commit with a new input file by running the relevant `yarn` commands to generate the output files, and commit the results to the same branch.
    - Pros: Everything tracked in git today is still tracked, closely mirrors the existing process.
    - Cons: Could be tricky to get the right trigger set up – we'd want to avoid having a workflow attempt committing to a protected branch. Something that triggers on a commit that modifies the relevant input file on any branch that is NOT `_staging` or `main` might suffice, but there could be a better way.
- Add the `yarn` commands as a post-checkout build step in the [Build Docker image for GOST API](https://github.com/usdigitalresponse/usdr-gost/blob/_staging/.github/workflows/build-api.yml) workflow so that the output files are bundled into the built Docker image artifact.
    - Pros: Doesn't require a workflow that modifies source code, triggers are straightforward as it would just get incorporated into every Docker build.
    - Cons: Could obscure the resulting output files and potentially make it harder to review/debug, as you'd need to pull the Docker image to inspect them.

I still think that moving away from this being a dev workflow and towards a feature of the app makes sense, but this is a possible interim MVP that would still streamline things more than they are today.

**Comment by mhuang-usdr** [comment]
I might be misunderstanding something, but feature to update the input template already exists.

![image](https://user-images.githubusercontent.com/91221265/235492290-807a0021-f6bd-49c6-871e-a782232aac1a.png)

**Comment by as1729** [comment]
@mhuang-usdr I believe both these processes update different templates. fwiw this ticket is intended to address the need for an engineer to update an input template which then automatically updates validation rules.

https://github.com/usdigitalresponse/usdr-gost/blob/_staging/docs/common-tasks/updating-templates.md#updating-input-templates---specifics

**Comment by as1729** [comment]
Note:
- there is already a manual workaround available - as long as we are able to support our partners at the end of the reporting cycle to upload these files (either in app or manually), then we should be fine taking a bit longer to complete this ticket

**Comment by as1729** [comment]
@as1729 to come up with a risk mitigation plan for this ticket.

**Comment by cajinajr** [comment]
Some updated wireframes I wanted to post/share after applying feedback from the first round ([Link to Figma](https://www.figma.com/file/jop0VhnUz2WgR9Ow8YJLaN/ARPA---Input-Validation-Template-Management?type=design&node-id=37%3A73&t=iPwcMk50iyjoGxJi-1))

Because two separate files are generated from the one excel upload, I thought it would be useful to mimic what can be seen in Github (thanks Aditya for directing me to that screen). This does complicate the design a bit, and I'd be curious to hear others' thoughts on the value of including this view, but I'm including it in the interest of making this a comprehensive workflow as outlined in the ticket.

![Input-Validation-Template--View](https://github.com/usdigitalresponse/usdr-gost/assets/16214088/8f488d29-c4f7-4c6e-ab0a-4f709023ba24)
Metadata such as _filename_, _updated on_ timestamp, and _updated by_ are still displayed for quick confirmation. A "restore previous" button is available to roll-back template versions just in case.

Below are two different versions of the screen that displays the changes made by the new upload. Instead of showing code, I tried listing out the changes in plain language to make it easier to understand what was happening.

The first is a _very_ simplified version:
![Input-Validation-Template--Changes1](https://github.com/usdigitalresponse/usdr-gost/assets/16214088/e0ded95c-a22a-4dcf-9e1a-562272d94a35)

The second version breaks up the changes by the two files that are generated:
![Input-Validation-Template-Changes2](https://github.com/usdigitalresponse/usdr-gost/assets/16214088/e00ffcb4-76f2-48c4-b6dd-2743d04ded3f)

I thought a confirmation modal that listed the total number of changes being made could be useful for introducing a bit of friction in the process. 
![Input-Validation-Template--Confirm-Friction](https://github.com/usdigitalresponse/usdr-gost/assets/16214088/ae1bc734-6790-47ab-b040-e5b032a6c7da)
There was a comment about auto-failing the process if the total number of changes exceeded a set limit, but I don't agree with that approach unless we were to communicate that to the user beforehand.

And finally, an exploration into what a live QA experience might look like in the process. Again, introduces a bit more complexity but could be useful in a future version of the workflow. 
![Input-Validation-Template--Live-QA-Exploration](https://github.com/usdigitalresponse/usdr-gost/assets/16214088/d077e2bd-bd0d-49fb-9dbe-dee867fd50b4)

Overall a balancing act between keeping the process simple but not too simple and mitigating risk. Curious to hear thoughts from @as1729 @NatHillardUSDR (y'all wrote the tickets for input and output templates) and @jcomeau1. Happy to field questions and comments via Slack or GH as I continue to work on this.

**Comment by ClaireValdivia** [comment]
This is a nice to have as it would free up eng time. next step would be refresh of Josh's ux proposal above to assess the lift and if we think it is worth pursuing
